### ISSUE TITLE
cadvisor collects lots of unneeded systemd cgroups stats for kubelet

### ISSUE THREAD

**Issue by chenchun** [google/cadvisor#2115]
ref https://github.com/kubernetes/kubernetes/issues/71367.
I'm still thinking of how to let kubelet not collecting the arbitrary cgroups such as `/system.slice/NetworkManager.service`, `/system.slice/sshd.service`, etc.
One problem is that container runtime may not starts before kubelet and runtime cgroup path can be created after kubelet started, so it would be hard to add a static white list path prefix on creating cadvisor interface.
I'm thinking of two solutions. The first is to change cadvisor to make it watch the input cgroup paths dynamically when invoking `GetContainerInfoV2()` or `GetContainerInfo()` if it doesn't watch them before, so that rawContainerCgroupPathPrefixWhiteList isn't required before creating cadvisor interface.
The second is to add a blacklist to cadvisor raw factory. Blacklist has a heigher priority than rawContainerCgroupPathPrefixWhiteList, if the cgroup path is in the blacklist, raw factory denies to watch it whether or not it matches rawContainerCgroupPathPrefixWhiteList. Any suggestion?

**Comment by dashpole** [comment]
You could, with the current approach, place the kubelet and runtime in different cgroups (not within system.slice) using the `runtime-cgroups` and `kubelet-cgroups` flags.

**Comment by WanLinghao** [comment]
@dashpole  I don't think we could achieve what @chenchun needs with the current approach. What he wants is stop  collecting system raw containers information like /system.slice/sshd.service. 
And no matter where  kubelet or runtime cgroup exist, the cAdvisor will collect every child cgroup data in root cgroups.  It will not change by where kubelet or runtime cgroup exist. 
I think if we want to stop collecting system raw containers. We could set `docker_only=true&& rawPrefixWhiteList = "/system.slice/kubelet.service, /system.slice/docker.service.."`. This is just a raw plan but I want to confirm if this is what we want at first place.

**Comment by chenchun** [comment]
> @dashpole I don't think we could achieve what @chenchun needs with the current approach

@WanLinghao Yes, I'm quite confused about the current approach. How does it make to stop kubelet collects `/system.slice/sshd.service`?

> We could set docker_only=true&& rawPrefixWhiteList = "/system.slice/kubelet.service, /system.slice/docker.service.."

`docker_only` flag will be removed, but we can use this for now. I wonder if there is any replaceable flag?

> We could set docker_only=true&& rawPrefixWhiteList = "/system.slice/kubelet.service, /system.slice/docker.service.."

Like I said, one problem is that container runtime may not starts before kubelet and runtime cgroup path can be created after kubelet started, so it would be hard to add a static white list path prefix on creating cadvisor interface.

So I'm suggesting the two solutions, but it seems the second one of `adding a blacklist to cadvisor raw factory` will work unless you change your kubelet and runtime cgroup path out of system.slice and changing cgroup path is painful for existing clusters.

>  change cadvisor to make it watch the input cgroup paths dynamically when invoking GetContainerInfoV2() or GetContainerInfo() if it doesn't watch them before

I prefer this approach, is it acceptable?

**Comment by dashpole** [comment]
By the "current approach" I mean using [rawContainerCgroupPathPrefixWhiteList](https://github.com/google/cadvisor/blob/master/manager/manager.go#L145), rather than adding a new whitelist.  That one collects all cgroups within the provided paths.  So if you had the structure `/kubelet`, `/runtime`, `/kubepods/...`, etc., we can pass those in when we start cAdvisor and get only the metrics you care about.

**Comment by chenchun** [comment]
ref https://github.com/kubernetes/kubernetes/pull/21337 , kubelet does checking runtime cgroups periodically rather than passing those in when we start cAdvisor to avoids race with runtime uptime. So I don't think this is an acceptable approach.

**Comment by dims** [comment]
Closing this issue as it has been open since before January 1, 2024. If this is still relevant, please feel free to reopen or create a new issue.
